cooltools/lib/checks.py

```diff
@@ -0,0 +1,502 @@
+import warnings
+from itertools import tee, starmap
+from operator import gt
+from copy import copy
+
+import numpy as np
+import pandas as pd
+import bioframe
+
+from . import schemas
+from .common import make_cooler_view
+
+def _is_sorted_ascending(iterable):
+    # code copied from "more_itertools" package
+    """Returns ``True`` if the items of iterable are in sorted order, and
+    ``False`` otherwise.
+
+    The function returns ``False`` after encountering the first out-of-order
+    item. If there are no out-of-order items, the iterable is exhausted.
+    """
+
+    it0, it1 = tee(iterable) # duplicate the iterator
+    next(it1, None) # skip 1st element in "it1" copy
+    # check if all values in iterable are in ascending order
+    # similar to all(array[:-1] < array[1:])
+    _pairs_out_of_order = starmap(gt, zip(it0, it1) )
+    # no pairs out of order returns True, i.e. iterator is sorted
+    return not any(_pairs_out_of_order)
+
+def is_expected(
```

Comment:
I'm not full convinced this one is useful for users - I had it not hidden at first, but then added `_` - because `is_compatible_expected(...verify_cooler=None...)` does roughly the same

I made it hidden to keep the total number of visible checks smaller, less confusing

what are the arguments for making it visible ?

Reply:
Aha, hadn't realized that base case could be assayed w/o passing a view. Maybe rename to is_valid_expected? Compatible makes me think it should strictly require two inputs. I also think I saw the _is_expected used outside of is_compatible_expected.

Reply:
it might have been used separately in the `read_expected_from_file` before, but now it shouldn't be used anywhere outside of `is_compatible_expected` (will check)

All for renaming !  `is_valid_expected` does make more sense indeed !
although I wanted to ask/discuss what do this check should return - shoud they return `True/False` and optionaly `raise`, or should they return `list` of "issues" with the input ... @nvictus had some suggestions - but in the rush before the release we kept them as is

Reply:
oh yeah, `_is_expected` is still used separately in `read_expected_from_file` - this can (and should) be changed

Reply:
Ok! I'll:
- roll back to _is_expected and
- go ahead with the other renaming
- update read_expected_from_file to only depend on is_valid_expected

Reply:
Sounds great !

Reply:
Running into an issue: it seems that is_compatible_expected will not accept None for the view. 

![image](https://user-images.githubusercontent.com/19657573/144922368-e7f0fb4f-fc76-4e24-bd0e-2353a41ebfb8.png)

I guess I can just make this execute if verify_view is not None: https://github.com/open2c/cooltools/blob/6c9f4ff8ae8f8dade313812fe8c9391d773ddb5d/cooltools/lib/checks.py#L207
?

Reply:
also-- shouldn't this be moved inside the for loop? https://github.com/open2c/cooltools/blob/6c9f4ff8ae8f8dade313812fe8c9391d773ddb5d/cooltools/lib/checks.py#L237

Reply:
oh yeah - it looks like `if n_diags_expected != n_diags_cooler: raise ...` should inside of the `for` loop over regions! indeed!
thank you !